test-data/unit/check-newsemanal.test

```diff
@@ -336,6 +312,81 @@ class A: pass # E: Name 'A' already defined (possibly by an import)
 class A: pass
 from a import x
 
+[case testNewAnalyzerClassInFunction]
+def main() -> None:
+    x: C
+    class C:
+        def __init__(self) -> None:
+            self.x: A
+            x()  # E: "C" not callable
+    reveal_type(x.x)  # E: Revealed type is '__main__.A@8'
+    class A: pass
+
+[case testNewAnalyzerMutuallyRecursiveFunctions]
+def main() -> None:
+    def f() -> int:
+        reveal_type(g())  # E: Revealed type is 'builtins.str'
+        return int()
+    def g() -> str:
+        reveal_type(f())  # E: Revealed type is 'builtins.int'
+        return str()
+
+[case testNewAnalyzerMissingNamesInFunctions]
+def main() -> None:
+    def f() -> None:
+        x  # E: Name 'x' is not defined
+        class C:
+            x  # E: Name 'x' is not defined
+
+[case testNewAnalyzerMutuallyRecursiveOverloadedFunctions]
+from typing import overload, Union
+
+def main() -> None:
+    @overload
+    def f(x: int) -> int: ...
+    @overload
+    def f(x: str) -> str: ...
+    def f(x: Union[int, str]) -> Union[int, str]:
+        reveal_type(g(str()))  # E: Revealed type is 'builtins.str'
+        return x
+    @overload
+    def g(x: int) -> int: ...
+    @overload
+    def g(x: str) -> str: ...
+    def g(x: Union[int, str]) -> Union[int, str]:
+        reveal_type(f(int()))  # E: Revealed type is 'builtins.int'
+        return float()  # E: Incompatible return value type (got "float", expected "Union[int, str]")
+
+[case testNewAnalyzerNestedClassInMethod]
+class C:
+    class D:
+        def meth(self) -> None:
+            x: Out.In
+            reveal_type(x.t)  # E: Revealed type is 'builtins.int'
+            class Out:
+                class In:
+                    def meth(self) -> None:
+                        self.t: int
+
+[case testNewAnalyzerDeeplyNestedFunctions]
+class Out:
+    class In:
+        def meth(self) -> None:
+            x: C.D
+            reveal_type(x.t)  # E: Revealed type is '__main__.Test@10'
+            class C:
+                class D:
+                    def meth(self) -> None:
+                        self.t: Test
+                        class Test:
+                            def test(self) -> None:
+                                def one() -> int:
+                                    reveal_type(other())  # E: Revealed type is 'builtins.str'
+                                    return int()
+                                def other() -> str:
+                                    reveal_type(one())  # E: Revealed type is 'builtins.int'
+                                    return str()
```

Comment:
It would be good to have a test case where we hit the maximum iteration count (cyclic definition).